File: BackEnd-Siukat/services/listrik.go
package services

import (
	"BackEnd-Siukat/config"
	"BackEnd-Siukat/models"
	"time"
)

type ListrikService struct{}

func (s *ListrikService) Add(req models.Listrik, atribut string) (models.Listrik, error) {
	req.Atribut = atribut
	err := config.DB.Create(&req).Error
	return req, err
}

func (s *ListrikService) Edit(req models.Listrik, noPeserta string, existing models.Listrik, atribut string) (models.Listrik, error) {
	data := map[string]interface{}{
		"no_pelanggan":    req.NoPelanggan,
		"jenis_pemakaian": req.JenisPemakaian,
		"pengeluaran":     req.Pengeluaran,
	}

	if req.ScanListrik != "" {
		data["scan_listrik"] = req.ScanListrik
	} else {
		data["scan_listrik"] = existing.ScanListrik
	}

	if err := config.DB.Model(&models.Listrik{}).Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).Updates(data).Error; err != nil {
		return models.Listrik{}, err
	}
	var res models.Listrik
	config.DB.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&res)
	return res, nil
}

func (s *ListrikService) AddLog(user models.Listrik, atribut string, executor string, timestamp *time.Time) error {
	logLst := models.LogListrik{
		NoPeserta:      user.NoPeserta,
		NoPelanggan:    user.NoPelanggan,
		JenisPemakaian: user.JenisPemakaian,
		Pengeluaran:    user.Pengeluaran,
		ScanListrik:    user.ScanListrik,
		Atribut:        atribut,
		Executor:       executor,
		Timestamp:      timestamp,
	}
	return config.DB.Create(&logLst).Error
}

func (s *ListrikService) GetByLoggedIn(noPeserta string) (models.Listrik, error) {
	var res models.Listrik
	err := config.DB.Where("no_peserta = ?", noPeserta).First(&res).Error
	return res, err
}

func (s *ListrikService) CheckData(noPeserta string) (bool, error) {
	var res models.Listrik
	if err := config.DB.Where("no_peserta = ?", noPeserta).First(&res).Error; err != nil {
		return false, err
	}

	if res.NoPelanggan == "" || res.JenisPemakaian == "" || res.Pengeluaran == 0 || res.ScanListrik == "" {
		return false, nil
	}
	return true, nil
}

File: ZhammDataBeLama/siukat-service-dev_snmptn_2022/services/listrik.js
var Listrik = require("../models/listrik");
var LogListrik = require("../models/log_listrik");

var listrik = function () {};

listrik.prototype.add = function (req, atribut) {
  return new Promise(function (resolve, reject) {
    Listrik.create(
      {
        no_peserta: req.no_peserta,
        no_pelanggan: req.no_pelanggan,
        jenis_pemakaian: req.jenis_pemakaian,
        pengeluaran: req.pengeluaran,
        scan_listrik: req.scan_listrik,
        atribut: atribut,
      },
      {
        fields: [
          "no_peserta",
          "no_pelanggan",
          "jenis_pemakaian",
          "pengeluaran",
          "scan_listrik",
          "atribut",
        ],
      }
    )
      .then(function (response) {
        resolve(response);
      })
      .catch(function (err) {
        reject(err);
      });
  });
};

listrik.prototype.edit = function (req, no_peserta, atribut) {
  return new Promise(function (resolve, reject) {
    var data = {
      no_pelanggan: req.body.no_pelanggan,
      jenis_pemakaian: req.body.jenis_pemakaian,
      pengeluaran: req.body.pengeluaran,
    };

    if (req.file !== undefined) {
      data.scan_listrik = req.file.filename;
    }

    Listrik.update(data, {
      where: {
        no_peserta: no_peserta,
        atribut: atribut,
      },
    })
      .then(function (response) {
        resolve(response);
      })
      .catch(function (err) {
        reject(err);
      });
  });
};

listrik.prototype.addLog = function (users, atribut, executor, timestamp) {
  return new Promise(function (resolve, reject) {
    LogListrik.create(
      {
        no_pelanggan: users.no_pelanggan,
        jenis_pemakaian: users.jenis_pemakaian,
        pengeluaran: users.pengeluaran,
        scan_listrik: users.scan_listrik,
        atribut: atribut,
        executor: executor,
        timestamp: timestamp,
      },
      {
        fields: [
          "no_peserta",
          "no_pelanggan",
          "jenis_pemakaian",
          "pengeluaran",
          "scan_listrik",
          "atribut",
          "executor",
          "timestamp",
        ],
      }
    )
      .then(function (response) {
        resolve(response);
      })
      .catch(function (err) {
        reject(err);
      });
  });
};

listrik.prototype.getByLoggedIn = function (no_peserta) {
  return new Promise(function (resolve, reject) {
    Listrik.findOne({
      where: {
        no_peserta: no_peserta,
      },
    })
      .then(function (response) {
        resolve(response);
      })
      .catch(function (err) {
        reject(err);
      });
  });
};

listrik.prototype.checkData = function (no_peserta) {
  return new Promise((resolve, reject) => {
    Listrik.findOne({
      where: {
        no_peserta: no_peserta,
      },
      raw: true,
    })
      .then(function (response) {
        var result = true;
        for (var key in response) {
          result *= response[key] !== null && response[key] !== "";
        }
        resolve(result);
      })
      .catch(function (err) {
        reject(err);
      });
  });
};

module.exports = listrik;
